### ISSUE TITLE
Landmark problem due to insufficient precision

### ISSUE THREAD

**Issue by karussell** [graphhopper/graphhopper#2104]
For the bidirectional landmark algorithm (likely also for unidir LM) it can happen that the precision is not sufficient even for small areas and when the factor is properly estimated. Especially if roads are grid-like like in RoutingAlgorithmWithOSMTest.testCampoGrande where ShortestWeighting is used. In general a perfect-consistent weight estimate is not possible (as we store the weight as `short` and loose precision) and when there are multiple very similar alternatives for a route from A to B a problem can happen when the longer of those very similar alternatives is earlier explored than the shorter one (due to bigger rounding errors), which can lead to the problem that the SPTEntry is polled from the queue although the now shorter alternative wants to update it and `prioQueue.remove` returns false in AbstractNonCHBidirAlgo:

```
} else if (entry.getWeightOfVisitedPath() > weight) {
  boolean removed = prioQueue.remove(entry);
```

You can trigger a problem if you [throw an error in these cases](https://github.com/graphhopper/graphhopper/commit/d8554e943d0dd1b0a6f47698961061642a7676ca):

```
if (!removed)
  throw new IllegalArgumentException("Cannot find element with object " + entry + " in queue: " + prioQueue.toString());
```

It is similar to the fixed problem in #1749 (#1557) but I don't think this time we can or should do something about it.

That `removed` is `false` is not really a problem as the new, smaller entry is just added again to the priority queue even if the previous entry wasn't found. Also the weights of the alternatives must be very close and so it has no practical impact (<0.1m for ShortestWeighting). Even more seldom is the case that the slightly better alternative cannot be found (if the search stops too early) and we have not yet seen this case in tests otherwise landmark would return a different path than e.g. Dijkstra.

Still I stumbled over this problem when I tried to implement a slightly more efficient `prioQueue.update` method that requires the existence of the previous element in the queue. And it took me a long time to figure out what the real problem is. Increasing landmark-weight precision would improve the situation, but to guarantee that the `removed` variable is under all conditions `true` we would have to store the landmark weights lossless as `double` (8 bytes) I think, which we can't do memory-wise (not even 3 byte instead of 2).

I thought I still have to document this as an issue.

**Comment by easbar** [comment]
Thanks for figuring this out and documenting it as well. I am not sure if I understand yet. Why are rounded weights related to `prioQueue.remove`? `PriorityQueue#remove` uses `indexOf` so it tries to find an element that equals the given one. Since `SPTEntry` does not override `equals` this means we are looking at object equality `==`. However, I am not 100% sure how this interacts with the `SPTEntry#compareTo` implementation. Is object equality based on the weights because `SPTEntry` implements `compareTo`? Would it help to implement `equals` (using the `adjNode`). And if you use another type of queue could this not be done in a way that we remove (or update) the element by `adjNode` as well? We do not care about elements with similar weights afterall. All we want to do is update the element with a certain traversalId (adjNode for node-based).

**Comment by karussell** [comment]
Yes, for PriorityQueue we do not use rounded weights and all is fine there. The rounding issue happens at an earlier level or maybe I found a wrong explanation :)

What we need for A* is a heuristic estimate that underestimates the distance. This is fine even with rounding issues as we always "round down" for landmarks. But the problem is IMO that the heuristic needs to be consistent, i.e. it should always "round down" in the same manner but it does not in our case as the rounding errors can fluctuate between 0 and 1. And so for a slightly longer alternative to the destination it could happen that there is a constellation where the landmark weights are "rounded" towards picking a slightly longer alternative and explore it slightly ahead of the real shortest path. When this happens at some point the SPTEntry of the slightly longer alternative could be polled from the PriorityQueue before the shortest path tree explores the same SPTEntry for the correct shortest path and then it tries to update the existing SPTEntry but it won't find it in the PriorityQueue anymore, i.e. `removed` is `false`. The chances for that are very very low and even if there are such problems the differences are often in the area of <1e-10 but especially for the previously mentioned testCampoGrande I saw "big" differences of nearly 0.1m.

This problem [can also happen without landmarks](https://github.com/graphhopper/graphhopper/blob/master/core/src/main/java/com/graphhopper/routing/AStar.java#L114) I think, but the differences are likely always <1e-10.
